p2p/src/util/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+        let ident = if let Ok(syn::Meta::NameValue(syn::MetaNameValue {
+            lit: syn::Lit::Str(litstr),
+            ..
+        })) = attr.parse_meta()
+        {
+            syn::Ident::new(litstr.value().as_str(), Span::call_site())
+        } else {
+            panic!("`name` attribute in invalid form");
+        };
+
+        return Some(quote! { #ident });
+    }
+
+    None
+}
+
+// try to find entry with `#[return_value]` attribute and extract its type into `proc_macro2` token stream
+fn parse_return_value(
+    fields: &Fields,
+) -> Option<(proc_macro2::TokenStream, proc_macro2::TokenStream)> {
+    let retvals = fields
+        .iter()
+        .flat_map(|field| {
+            field
+                .attrs
+                .iter()
+                .find(|attr| attr.path.is_ident("return_value"))
+                .map(|_| field)
+        })
+        .collect::<Vec<_>>();
+
+    assert!(retvals.len() <= 1, "only one return value allowed");
+
+    retvals.get(0).map(|field| {
+        let ty = &field.ty;
+        let ident = &field.ident;
+        let retval: Retval = parse2(ty.into_token_stream())
+            .expect("Return value is expected to be in form `oneshot::Sender<type>`");
+        let inner = retval.inner;
+
+        (
+            quote! {
+                #ident
+            },
+            quote! {
+                #inner
+            },
+        )
+    })
+}
+
+#[proc_macro_derive(Handle, attributes(name, return_value))]
+pub fn event(input: TokenStream) -> TokenStream {
+    let data = parse_macro_input!(input as DeriveInput);
+    let orig = data.ident.clone();
+    let sender = syn::Ident::new(format!("{}Sender", data.ident).as_str(), Span::call_site());
+
+    let filtered = if let syn::Data::Enum(syn::DataEnum { variants, .. }) = data.data {
+        variants
+            .into_pairs()
+            .filter_map(|x| match x {
+                Punctuated(entry, _) => Some(entry),
+                _ => None,
+            })
+            .collect::<Vec<_>>()
+    } else {
+        panic!("`#[derive(Handle)]` can only be used with enums!");
+    };
+
+    let methods = filtered
+        .iter()
+        .map(|f| {
+            let name = &f.ident;
+            let fields = &f.fields;
+
+            let real_name = parse_real_name(&f.attrs).unwrap_or_else(|| {
+                let name = syn::Ident::new(&name.to_string().to_case(Case::Snake), Span::call_site());
+                quote! { #name }
+            });
+
+            let (func_args, chan_args): (Vec<_>, Vec<_>) = fields
+                .iter()
+                .map(|field| {
+                    let ident = &field.ident;
+                    let ty = &field.ty;
+
+                    (!field
+                        .attrs
+                        .iter()
+                        .any(|attr| attr.path.is_ident("return_value")))
+                    .then_some((
+                        quote! {
+                            , #ident: #ty
```

Comment:
This is a minor style issue but I'd prefer the fragments here to just expand to `#ident: #ty` (without the comma) and generate the commas when it's spliced into the code later `#(#func_args)*` -> `#(, #func_args)*`.